pkg/kv/postgres/store.go

```diff
@@ -0,0 +1,237 @@
+package postgres
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+	"github.com/treeverse/lakefs/pkg/kv"
+)
+
+type Driver struct{}
+
+type Store struct {
+	Pool   *pgxpool.Pool
+	Params *Params
+	closed bool
+}
+
+type Entries struct {
+	rows  pgx.Rows
+	entry *kv.Entry
+	err   error
+}
+
+const (
+	DriverName = "postgres"
+
+	defaultTableName = "kv"
+	paramTableName   = "lakefskv_table"
+)
+
+//nolint:gochecknoinits
+func init() {
+	kv.Register(DriverName, &Driver{})
+}
+
+func (d *Driver) Open(ctx context.Context, name string) (kv.Store, error) {
+	// TODO(barak): should we handle Open reuse the same store based on name
+	config, err := pgxpool.ParseConfig(name)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", kv.ErrDriverConfiguration, err)
+	}
+	pool, err := pgxpool.ConnectConfig(ctx, config)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", kv.ErrConnectFailed, err)
+	}
+	defer func() {
+		// if we return before store uses the pool, free it
+		if pool != nil {
+			pool.Close()
+		}
+	}()
+
+	// acquire connection and make sure we reach the database
+	conn, err := pool.Acquire(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", kv.ErrConnectFailed, err)
+	}
+	defer conn.Release()
+	err = conn.Conn().Ping(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", kv.ErrConnectFailed, err)
+	}
+
+	params := parseStoreConfig(config.ConnConfig.RuntimeParams)
+
+	err = setupKeyValueDatabase(ctx, conn, params)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", kv.ErrSetupFailed, err)
+	}
+	store := &Store{
+		Pool:   pool,
+		Params: params,
+	}
+	pool = nil
+	return store, nil
+}
+
+type Params struct {
+	TableName string
+}
+
+func parseStoreConfig(runtimeParams map[string]string) *Params {
+	p := &Params{
+		TableName: defaultTableName,
+	}
+	if tableName, ok := runtimeParams[paramTableName]; ok {
+		p.TableName = tableName
+	}
+	return p
+}
+
+// setupKeyValueDatabase setup everything required to enable kv over postgres
+func setupKeyValueDatabase(ctx context.Context, conn *pgxpool.Conn, params *Params) error {
+	_, err := conn.Exec(ctx, `CREATE TABLE IF NOT EXISTS `+params.TableName+` (
+    key BYTEA NOT NULL PRIMARY KEY,
+    value BYTEA NOT NULL);`)
+	return err
+}
+
+func (s *Store) Get(ctx context.Context, key []byte) ([]byte, error) {
+	if key == nil {
+		return nil, kv.ErrMissingKey
+	}
+	row := s.Pool.QueryRow(ctx, `SELECT value FROM `+s.Params.TableName+` WHERE key = $1`, key)
```

Comment:
This is vaguely unsafe -- the pq  drivers allow quoting identifiers, and this might be a way to inject SQL in some situations (if I can control just the table name, I can pwn the database after this line).

Reply:
Will look for a way to do it for a table name - not like query variables table names can't be set with placeholders. The only solution I found to help was to quote it.
Currently the table name can be altered just by configuration and nothing from user data/input.

Reply:
Updated the code to use sanitize table name.